src/libraries/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.Shared.xml

```diff
@@ -51,4 +51,35 @@
       <attribute internal="RemoveAttributeInstances" />
     </type>
   </assembly>
+
+  <!-- COM specific attributes -->
+  <assembly fullname="System.Private.CoreLib" feature="System.Runtime.InteropServices.Marshal.IsComSupported" featurevalue="false">
+    <type fullname="System.MTAThreadAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
+    <type fullname="System.STAThreadAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
+    <type fullname="System.Runtime.InteropServices.ClassInterfaceAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
+    <type fullname="System.Runtime.InteropServices.CoClassAttribute">
```

Comment:
Is stripping these attributes buying us anything? If user chooses to turn off COM, but they still have a bunch of COM classes that are kept, wouldn't it be better if they get useful feedback about it?

Stripping these attributes results in obnoxious failure modes like "System.Security.SecurityException: 'ECall methods must be packaged into a system module.'".

(That's the exception you get if you try to new up a class like this: `class Foo { [MethodImpl(MethodImplOptions.InternalCall)] public extern Foo(); }`, which is what's left of COM after these are stripped.)

Reply:
> Is stripping these attributes buying us anything?

It saves some size but it also correctness fix for us that we don't have a dependency on these attribute which is not guarded under COM feature.

> wouldn't it be better if they get useful feedback about it?

They have no meaning for example for MonoVM. Can CoreCLR throw a better error when the runtime starts in COM disabled mode?

Reply:
> It saves some size but it also correctness fix for us that we don't have a dependency on these attribute which is not guarded under COM feature.

Wouldn't it be better to keep these so that output can be inspected for these? I assume if we have a class marked with `CoClassAttribute`, we would want to remove _the entire class_ because any code calling into it is going to crash with the useless exception I quoted above. That's a much better size saving and also a correctness fix because instead of producing reachable code that might crash at runtime, we can prove the crashing code is not there. Kind of feels like COM stripping should be something where linker is more involved and can issue warnings ("hey, I disabled COM, but you still have COM CoClasses left, things may crash at runtime").

> Can CoreCLR throw a better error when the runtime starts in COM disabled mode?

The runtime doesn't know this was a COM class if linker strips all signs of it. It's just a class with an `InternalCall` constructors that the runtime doesn't know what to do with.

Reply:
> I assume if we have a class marked with CoClassAttribute, we would want to remove the entire class because any code calling into it is going to crash with the useless exception I quoted above.

I don't follow. You apply CoClassAttribute to an interface which type argument of the implementation class. By removing the attribute from the interface you remove the class if there are no further references. 

> The runtime doesn't know this was a COM class if linker strips all signs of it

It could, this is standard opt-in feature switch so when someone disables it on Windows you will have the information written in *.runtimeconfig.json

/cc @vitek-karas

Reply:
> I don't follow. You apply CoClassAttribute to an interface which type argument of the implementation class. By removing the attribute from the interface you remove the class if there are no further references.

Okay, let's say I referred to/commented on the ComImportAttribute a couple lines below. I didn't have my second coffee yet.

Reply:
I think we'll have to leave the MTA/STA alone - those have effect outside of COM support - and even if the app itself doesn't use COM, it still may need to set STA for other Windows stuff to work correctly.